### PR TITLE
feature(core): Add coalescing support to lookup v2

### PR DESCRIPTION
**Pull request by fuchsnj**
This adds coalescing support to lookup v2. With this change, lookup v2 has full feature parity with the previous lookup implementation and can start replacing it.

The micro benchmarks for parsing are nearly identical to before, even with the added features. It is roughly 2000x faster than the current larlpop based parser.

changes:
- add coalesce parsing support to lookup v2
- add coalesce support to all CRUD `value` functions.

### CODE REVIEW

**Review by pablosichert — APPROVED** [review]
Looks good!

Some questions, but not blocking:

File: lib/value/src/value/crud/insert.rs [hunk 1]
```diff
@@ -46,6 +68,53 @@ mod test {
...
+        value.insert("(c|a)", 13);
+        assert_eq!(
+            value,
+            Value::from(json!({
+                "a": 13,
+                "b": {"x": true}
+            }))
+        );
+
+        value.insert("(a|b).x", 5);
```
[the remark this answers is not in the document]
Comment (fuchsnj) [reply lib/value/src/value/crud/insert.rs:108]: This is something we will probably prevent with a compile-time error, but the functionality itself won't go away. (For example Vector itself could still do this, but not a user through VRL)

File: lib/value/src/value/crud/remove.rs [hunk 2]
```diff
@@ -34,6 +42,31 @@ mod test {
     use crate::Value;
     use serde_json::json;
 
+    #[test]
+    fn remove_coalesce() {
+        let mut value = Value::from(json!({
+            "field": 123,
+            "field2": 321
+        }));
+        assert_eq!(value.remove("(foo|bar)", false), None);
```
[the remark this answers is not in the document]
Comment (fuchsnj) [reply lib/value/src/value/crud/remove.rs:51]: 2nd param is `prune`, which removes empty collections if they are empty as the result of the removal. This is tested in the `remove_prune` test below (coalescing doesn't really change this)

File: lib/lookup/src/lookup_v2/owned.rs [hunk 3]
```diff
@@ -96,6 +99,34 @@ impl Serialize for OwnedPath {
     }
 }
 
+fn serialize_field(field: &str, separator: Option<&str>) -> String {
+    let needs_quotes = field
+        .chars()
+        .any(|c| !matches!(c, 'A'..='Z' | 'a'..='z' | '_' | '0'..='9' | '@'));
```
Comment (JeanMertz) [inline lib/lookup/src/lookup_v2/owned.rs:105]: Might be good to add a comment here, pointing to the VRL parser, to maker sure we keep the list of allowed characters in sync.

File: lib/lookup/src/lookup_v2/owned.rs [hunk 4]
```diff
@@ -238,6 +282,11 @@ mod test {
             (r#"foo."a\"a"."b\\b".bar"#, r#"foo."a\"a"."b\\b".bar"#),
             ("<invalid>", "<invalid>"),
             (r#""🤖""#, r#""🤖""#),
+            (".(a|b)", "(a|b)"),
+            (".(a|b|c)", "(a|b|c)"),
+            ("foo.(a|b|c)", "foo.(a|b|c)"),
+            ("[0].(a|b|c)", "[0].(a|b|c)"),
+            (".(a|b|c).foo", "(a|b|c).foo"),
```
Comment (JeanMertz) [inline lib/lookup/src/lookup_v2/owned.rs:289]: Probably a good idea to add a few tests with whitespacing, to ensure that is parsed as expected.

**Review by JeanMertz — APPROVED** [review]
Sweet. A few nits, but otherwise LGTM.

File: lib/lookup/src/lookup_v2/owned.rs (hunk 4 above)
Comment (fuchsnj) [reply lib/lookup/src/lookup_v2/owned.rs:289]: These are the serialization tests. There is a whitespace test on the parser already. I'll add another here though.

**Pull request merged**